text/0000-orderer-snapshot.md

```diff
@@ -0,0 +1,316 @@
+2. BootstrappingSnapshotInfo
+
+For BootstrappingSnapshotInfo persistence the implementation is already there, but for the last config block a new persistent unit of storage (a file in case of file-based ledger) should be created.
+A special attention should be paid to the use case of **Deliver** service being asked to deliver a single block with the number equal to the block number of the last config block from a snapshot - in this case the delivery request must succeed by returning the last config block from the new persistent unit of storage (from a separate file) and not from the continuous block ledger. In general, the ledger storage interface should be aware of a possibility that aside from a continuous block ledger there is a special case of a single block representing a last config block from a snapshot which should also be successfully returned when requested by block number.
+Additionally all the procedures inside the OSN codebase looking up the last config block in the OSN ledger storage should be modified to return the last config block from the snapshot in case the last config block and/or a last block is not found in the block ledger itself.
+
+The channel bootstrapping from snapshot must be crash-consistent, that is if a crash takes place while bootstrapping a channel, upon restart the orderer, either the bootstrapped channel exists in the usable state or not at all. In particular this means that the creation of the above two pieces of information (last config block and BootstrappingSnapshotInfo) must be an atomic transaction, and there should be no possibility after a crash to have one of them but not another.
+Possible implementations of such an atomic transaction may be keeping both pieces of information in the same file and creating the file atomically, or keeping them in separate files in the same directory and creating the directory atomically.
+
+## System channel co-existence
+
+The following API can work both with system channel and without it:
+- "channel list" - existing API, keep existing behaviour
+
+The following APIs can work only without system channel, will fail in presence of system channel:
+- "channel joinbysnapshot"
+- "channel remove" - existing API, keep existing behaviour
+- "channel join" - existing API, keep existing behaviour
+- "ledger snapshot"
+
+## New channel creation ("genesis" snapshot)
+
+A new channel can be created using the existing API "channel join" with a genesis block as a join-block.
+
+Theoretically a "genesis" snapshot also can be constructed. For this a genesis block should be used as a last config block and lastBlockNum should be set to 0. This RFC does not define any specific tool capable of creation of a "genesis" snapshot, instead this RFC mandates that such a snapshot should be accepted as a valid one. In particular a snapshot should be considered as valid if prevBlockHash is omitted from the metadata JSON file or is set to "null", but only when lastBlockNum is equal to 0.
```

Comment:
what's so special about this genesis snapshot? why do we need it?

Reply:
This is a nice feature to have. We do not strictly need it - as explained in the text, "channel join" API can be used to implement the same functionality.
But it makes the snapshot API self-sufficient and allows admins to operate with snapshots without worrying about an exceptional case of block number 0. And the implementation of it should be trivial.